### ISSUE TITLE
Cannot infer type of class variables when they're created via instance methods

### ISSUE THREAD

**Issue by spindlebink** [crystal-lang/crystal#11805]
I'm unsure if this is by design, but it looks like class variables don't get type-inferred when they're initialized using instance methods. Although you'll probably never be defining instance variables based on instances from most classes, this issue does preclude class variables based on methods that work on primitives. For example:
```crystal
class Thing
  DELAY = 10.milliseconds

  @timer_countdown = DELAY # can't infer the type of instance variable '@timer_countdown' of Thing
end
```

**Comment by asterite** [comment]
Right, you have to say the type of the instance variable. This is by design.

**Comment by spindlebink** [comment]
It was my understanding that since instance methods like `Int.milliseconds` are defined [with an explicit return type of `Time::Span`](https://github.com/crystal-lang/crystal/blob/932f193ae/src/time/span.cr#L477), type inference would pick up that a call to `10.milliseconds` is restricted to that type. With most instance variable declarations, types get inferred as expected:
```
class Other
  def Other.get_the_thing(n) : Int32
    50 + n
  end
end

class Thing
  CONSTANT = 10

  @instance_bool = true
  @instance_other = Other.new
  @instance_span = Time::Span.new
  @instance_call = Other.get_the_thing 25
  @instance_constant = CONSTANT

  @instance_from_variable_method = 10.milliseconds # error as mentioned
end
```
It's just in this one case of calling a method on a variable (even a method with restricted type) that it doesn't work, which I see as a minor inconsistency--am I misunderstanding how type inference works here?

**Comment by asterite** [comment]
There are some special rules for what the compiler can infer for instance and class variables. It all boils down to simplicity and not having to trigger the full semantic phase to compute these things.

When the compiler sees `Foo.bar` it's very easy to determine: the method must be defined in `Foo` and there must be a method `bar`. If it has a return type, that's the type.

For more complex things like `a.foo` the compiler would need to first determine what's the type of `a`, and that's more complex.

That said, if it's a method on an integer literal, or actually any literal, the compiler could be a bit smarter, and we could consider adding that to the list of rules. In general I think the rule could be:
- If it's a call, and all arguments to the call are "simple", then try to deduce the type

The definition of "simple" above would be:
- A type literal (like `Foo`)
- Any other literal, like `1`, `'a'`, `"hello"`
- A call that's also "simple" (so the definition is actually recursive)

Well, actually, there are a bunch of rules for what is "simple" right now, so we only need to enhance the definition of what's a "simple call."

That would mean you could then use expressions like this to initialize instance variables without specifying an explicit type:
- `2 * 3`
- `2 ** 10`
- `"hello".upcase` (not very useful :-P)
- `10.milliseconds` (this one is useful)
- `'a'.ord`

But I'd say this is very low priority because the workaround is just writing `@timer_countdown : Time::Span = DELAY`

We also need to know what the opinion of others is regarding this change.
